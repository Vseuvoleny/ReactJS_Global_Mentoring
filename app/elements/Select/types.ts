import { IOption } from "./Option/types";

export interface ISelect extends IOption {
    placeholder: string,
    label: string;
    value: any[];
    errors: any;
    touched: any
}