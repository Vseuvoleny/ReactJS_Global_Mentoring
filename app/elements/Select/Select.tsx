import React, { FC, useState } from "react";
import RedTriangle from "../Icon/Icons/RedTriangle/RedTriangle";
import Option from "./Option/Option";
import classes from "./Select.module.scss";
import { ISelect } from "./types";

// TODO
// сделать закрытие через outsideclick

const Select: FC<ISelect> = ({
  placeholder,
  label,
  value,
  options,
  name,
  ...rest
}) => {
  const [isListOpen, setListOpen] = useState(false);
  const { errors, touched } = rest;

  return (
    <div className={classes.select_input}>
      <span className={classes.label}>{label}</span>
      <div
        className={classes.select_input__box}
        onClick={() => setListOpen(!isListOpen)}
      >
        {value.length > 0 ? value.join(", ") : placeholder}
        <div className={classes.expanded_list}>
          <span className={classes.expanded_list__icon}>
            <RedTriangle viewBox="0 0 10 7" width="10" height="7" fill="none" />
          </span>
        </div>
      </div>
      {errors && touched.genres && (
        <div className={classes.error_message}>{errors.genres}</div>
      )}
      {isListOpen && <Option options={options} name={name} />}
    </div>
  );
};

export default Select;
