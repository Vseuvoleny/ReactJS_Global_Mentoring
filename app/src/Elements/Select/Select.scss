.select-input {
  position: relative;
  z-index: 1;
  .label {
    display: inline-block;
    font-weight: 600;
    font-size: 16px;
    line-height: 20px;
    letter-spacing: 1px;
    text-transform: uppercase;
    color: #f65261;
    opacity: 0.8;
    margin-bottom: 13px;
  }
  &__box {
    position: relative;
    background: rgba(50, 50, 50, 0.948044);
    opacity: 0.8;
    border-radius: 4px;
    padding: 18px 0 18px 18px;

    .expanded-list {
      z-index: 20;
      &__icon {
        right: 19px;
        top: 50%;
        transform: translateY(-50%);
        position: absolute;
      }
    }
  }
  .options {
    width: 100%;
    background: rgba(35, 35, 35, 0.918051);
    box-shadow: 0px 0px 2px rgba(0, 0, 0, 0.1), 0px 2px 10px rgba(0, 0, 0, 0.1),
      0px 10px 20px rgba(0, 0, 0, 0.1), 0px 10px 50px rgba(0, 0, 0, 0.1);
    backdrop-filter: blur(27.1828px);
    border-radius: 4px;
    z-index: 1;
    position: absolute;
    padding: 12px 18px 24px;
    .option {
      cursor: pointer;
    }
    .option + .option {
      margin-top: 18px;
    }
  }
}
