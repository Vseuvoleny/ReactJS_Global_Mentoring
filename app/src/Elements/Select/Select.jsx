import React, { useState } from "react";
import PropTypes from "prop-types";
import RedTriangle from "../Icon/Icons/RedTriangle/RedTriangle";
import Option from "./Option/Option";
import { Field } from "formik";
import "./Select.scss";

// TODO
// сделать закрытие через outsideclick

const Select = ({ placeholder, label, values, options, name, ...rest }) => {
  const [isListOpen, setListOpen] = useState(false);

  return (
    <div className="select-input">
      <span className="label">{label}</span>
      <div className="select-input__box">
        {values.length > 0 ? values.join(", ") : placeholder}
        <div className="expanded-list">
          <span
            onClick={() => setListOpen(!isListOpen)}
            className="expanded-list__icon"
          >
            <RedTriangle viewBox="0 0 10 7" width="10" height="7" fill="none" />
          </span>
        </div>
      </div>
      {isListOpen && (
        <div className="options">
          <Field name={name}>
            {({ field }) =>
              options.map((option) => (
                <React.Fragment key={option.key}>
                  <input
                    type="checkbox"
                    id={option.title}
                    value={option.title}
                    checked={field.value.includes(option.title)}
                    {...field}
                    {...rest}
                  />
                  <label htmlFor={option.title}>{option.title}</label>
                </React.Fragment>
              ))
            }
          </Field>
        </div>
      )}
    </div>
  );
};

Select.propTypes = {
  placeholder: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
};

export default Select;
