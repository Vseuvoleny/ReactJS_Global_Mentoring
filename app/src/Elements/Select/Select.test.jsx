import React from "react";
import { render } from "@testing-library/react";
import Select from "./Select";
import userEvent from "@testing-library/user-event";
import { Formik } from "formik";

describe("Select component", () => {
  const fakeOptions = [{ id: 1, title: "title" }];
  test("should render Select component", () => {
    const { container, getByText } = render(
      <Select label={"text_label"} value={[]} />
    );
    expect(container).toMatchSnapshot();
    expect(getByText("text_label")).toBeInTheDocument();
  });
  test("should expanded list to be render by click", () => {
    let isListOpened = false;
    const mockListChanger = jest.fn(() => (isListOpened = !isListOpened));
    const { getByText } = render(
      <Formik>
        <Select label={"text_label"} value={[]} options={fakeOptions} />
      </Formik>
    );
    const labelElement = getByText("text_label");
    const expandListControl = labelElement.nextSibling;
    expandListControl.addEventListener("click", mockListChanger);
    expect(isListOpened).toBeFalsy();
    userEvent.click(expandListControl);
    expect(mockListChanger).toHaveBeenCalled();
    expect(isListOpened).toBeTruthy();
  });
});
