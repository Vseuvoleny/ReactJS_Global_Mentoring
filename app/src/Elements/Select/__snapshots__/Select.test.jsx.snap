// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render Select component 1`] = `
<div>
  <div
    class="select-input"
  >
    <span
      class="label"
    >
      text_label
    </span>
    <div
      class="select-input__box"
    >
      <div
        class="expanded-list"
      >
        <span
          class="expanded-list__icon"
        >
          <svg
            fill="none"
            height="7"
            role="img"
            viewBox="0 0 10 7"
            width="10"
          >
            <path
              clip-rule="evenodd"
              d="M1.17339 0C0.319169 0 -0.141788 1.00184 0.413939 1.65057L4.23746 6.11398C4.63642 6.57971 5.35674 6.57992 5.75597 6.11442L9.58401 1.65101C10.1403 1.0024 9.67943 0 8.82494 0H1.17339Z"
              fill="#F65261"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
